Initial XBR protocol implementation

om26er opened the pull request:
Implements the XBR Buyer and Seller

There are things that still don't work, we still have a `sample.js` file that needs to be removed before landing it.

oberstet commented:
I closed above PR, but added CI test coverage for using native binary payloads across JS and Py via https://github.com/crossbario/autobahn-js/blob/caa775adbdecfe835a294b721f990e558009f87f/test/test_binary.js#L21 - IOW: this "should work" (calling the market maker with native bytes). we could also add CI tests actually running a crossbarfx node with market maker .. more work though;)

oberstet commented:
to enforce using CBOR, a WAMP connection can be created like this:

```javascript
var config = {
    url: "ws://127.0.0.1:8080/ws",
    realm: "realm1",
    serializers: [new autobahn.serializer.CBORSerializer()]
};

var connection = new autobahn.Connection(config);
```

oberstet commented:
yeah, this looks good! basic class structure is there and uses the new dependencies for actual crypto. guess it is quite close to full interop with ABPy based XBR stuff .. awesome=)

om26er commented:
The Buyer and Seller both are implemented now (close to ABPy). Lets lands this now and iterate on it based on feedback.

om26er commented on lib/xbr/buyer.js:
```diff
@@ -0,0 +1,118 @@
+    this._keyPair = nacl.box.keyPair();
+};
+
+SimpleBuyer.prototype.start = function(session, consumerID, on_success, on_failure) {
```
@oberstet Need your input on this API signature. Currently it takes two callbacks `on_success` and `on_failure` as parameters because `start()` makes a call to `session.call()` and returns instantly, hence we can't really return the result as we do in the Python implementation.

Are you fine with that approach or do we want to use something different here (and in other places below).

oberstet replied:
`start` should return a proper promise (not take callback args), and then you can fire the promise with success or error later on ..

oberstet replied:
rgd new release: sure!

om26er commented:
OK, now moved from callback(s) approach to deferred as used by other parts of the library.

Note: Moved around small part of the library code into utils.js.

oberstet commented:
@om26er should we merge this or do you want to piggyback more onto? if you wanna merge, go ahead (merge yourself)! looks good

om26er commented:
Ok, landing this now. More fixes/changes to follow
